File: 169.多数元素.cs
/*
 * @lc app=leetcode.cn id=169 lang=csharp
 *
 * [169] 多数元素
 */

// @lc code=start
public partial class Solution
{
    public int MajorityElement(int[] nums)
    {
        int num = nums[0];
        int count = 1;
        for (int i = 1; i < nums.Length; i++)
        {
            int currentNum = nums[i];
            if (currentNum == num)
            {
                count++;
            }
            else
            {
                if (count == 0)
                {
                    count = 1;
                    num = currentNum;
                }
                else
                {
                    count--;
                }
            }
        }
        return num;
    }
}
// @lc code=end

File: src/169.多数元素.rs
/*
 * @lc app=leetcode.cn id=169 lang=rust
 *
 * [169] 多数元素
 */

// @lc code=start
impl Solution {
    pub fn majority_element(nums: Vec<i32>) -> i32 {
        let mut result = nums[0];
        let mut count = 1;
        for index in 1..nums.len() {
            let current_num = nums[index];
            if current_num == result {
                count += 1;
            } else {
                if count == 0 {
                    count = 1;
                    result = current_num;
                } else {
                    count -= 1;
                }
            }
        }
        return result;
    }
}
// @lc code=end